plugins/uefi-sbat/README.md

```diff
@@ -0,0 +1,124 @@
+---
+title: Plugin: UEFI SBAT
+---
+
+## Introduction
+
+SBAT is a generation number based revocation mechanism that is used as a successsor to dbx checksum
+blocklisting.
+
+See [https://github.com/rhboot/shim/blob/main/SBAT.md](the specification) for more details.
+
+## Firmware Format
+
+The daemon will decompress the cabinet archive and extract a signed PE/COFF firmware blob with a
+`.sbata` section containing the revocation CSV data.
+
+This plugin supports the following protocol ID:
+
+* `org.uefi.sbat`
+
+## GUID Generation
+
+These devices use the standard USB DeviceInstanceId values, e.g.
+
+* `USB\VENDOR_fedora` (only quirk)
+* `USB\VENDOR_fedora&VAR_SbatLevelRT`
+
+## Update Behavior
+
+The device version is calculated using the same heuristics as for the revocation PE file, which is
+complicated slightly by there being potentially multiple components (each with a version) in the
+payload, for instance, `shim`, `grub` and `sd-boot`.
+
+These revocation components may or may not exist in the ESP, and may not exist in the existing
+platform UEFI `SbatLevelRT` variable, and so it is impossible to create a set of child-devices
+for each item like we might do for a USB device for example.
+
+Additionally, assigning *a single version* is complicated further because per-vendor overrides may
+be present in the SBAT entries in either the `SbatLevelRT` key or the signed revocation file.
+
+To overcome these limitations (and to force a round peg in a square hole) we sum up each generation
+number contained in the SBAT sections, **including** the per-vendor value where provided.
+For example:
+
+A single component:
+
+    sbat,1
+    -> "version 1"
+
+Multiple components:
+
+    sbat,1
+    grub,4
+    -> "version 5"
+
+Multiple components, with a distro specific override:
+
+    sbat,1
+    grub,4
+    grub.fedora,2
+    -> "version 7"
+
+This ensures that we can have a single version number representing the device current state, and
+also a single version number that represents the most up to date **per-distro** revocation data.
+
+For this reason different distributions should have a different AppStream namespace for the
+revocation data, for instance `org.fedoraproject.sbat.firmware` -- which makes sense as each vendor
+will be signing the revocation binary with its own key anyway.
```

Comment:
What happens with dual boot though? If you have 5 from Ubuntu and 5 from Fedora your version is 10. But you would have 6 on LVFS for Fedora and 6 from Ubuntu.

Reply:
I don't know: The SbatLevel mechanism isn't terribly well set up right now to support dual boot -- it's kinda designed on the assumption you only have one entity supplying `shimx64.efi` on a given system. Right now what happens is that distro 1 updates the shim package, which carries the revocation policy in it's own `.sbata` section. When distro 1 shim.efi gets run it automatically updates the SbatLevelRT key -- which would in theory block the older shim.efi from distro 2 being allowed to run. i.e. the shim that gets run first in BootOrder "wins".

I don't think it's possible to have two things writing to SbatLevelRT and so if you have an installed sbat of:

    sbat,1
    grub,4
    grub.fedora,2
    grub.ubuntu,2

and then Fedora offers you an update of

    sbat,1
    grub,4
    grub.fedora,3

I think it's correct that we don't try to deploy the update -- the installed version wasn't installed by Fedora and so we shouldn't try to update it. That said, maybe we do need to be more careful of the "non-distro" part, e.g. if we bump the `grub` generation by 1 then that's a way more important but than Ubuntu bumping the `grub.ubuntu` up to 4. Maybe we add up all the generations as before, but "generic" components like `grub` get bitshifted left a few places before adding.

Reply:
I guess in my mind what would be ideal is if there was a central authority distributing the SBAT revocations and all the distros used the common one.  So then fwupd can scan the ESP and just like DBX we only offer the SBAT revocation when there are no affected binaries there.

Reply:
A guess the logic for not centralizing it was that the revocation.efi binary has to be signed by the distro cert, not the MS cert. I don't know how popular it would be to use a RH-signed cert in Debian for example.